dts: pinctrl: Introduce pincfg-node and convert Atmel SAM as example

nandojve opened the pull request:
Many data items that are represented in a pin configuration node are common and generic.  Pin control bindings should use the properties defined on a standard way when are applicable; not all of these properties are relevant or useful for all hardware or binding structures.  Each individual binding document should state which of these generic properties, if any, are used, and the structure of the DT nodes that contain these properties. This intents to have a standard name in devicetree and be closer to Linux to easy understanding.

This is based on Linux, documentation:
https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml

The Atmel SAM was converted to use pincfg-node schema as an example for other follows.
Since this is platform dependent the best path is introduce the common schema and than migrate platforms when appropriated.

nandojve committed:
dts: pinctrl: Introduce pincfg-node

Many data items that are represented in a pin configuration node are
common and generic.  Pin control bindings should use the properties
defined on a standard way when are applicable; not all of these
properties are relevant or useful for all hardware or binding
structures.  Each individual binding document should state which of
these generic properties, if any, are used, and the structure of the
DT nodes that contain these properties.

This is based on Linux, documentation:
https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
dts: pinctrl: atmel: sam-pinctrl: Switch to pincfg-node

Add pincfg-node compatibility.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

gmarull reviewed (commented):
LGTM, one comment:
stm32 defines slew-rate as a string enum, will it be overriden correctly? ref. https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/pinctrl/st%2Cstm32-pinctrl.yaml#L76.
I understand that STM32 can move back to `int` if it ends up being a problem.

nandojve commented:
> LGTM, one comment:
> stm32 defines slew-rate as a string enum, will it be overriden correctly? ref. https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/pinctrl/st%2Cstm32-pinctrl.yaml#L76.
> I understand that STM32 can move back to `int` if it ends up being a problem.

Zephyr will expand all names that were declared when running dts preprocessor. I mean, if you move the enum names to STM32 include files, that are included in the dts preprocessor, that it will be expanded. Inclusive, the #28779 was motivated because of this behavior.

In my understanding, names in the devicetree files that allows users have a better understanding about the property can be expanded and should be used. However, at driver level, you won't be able to use the **DT_ENUM_*** as limitation, in this case.

Maybe @galak can help us to get the best view on this topic.

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -0,0 +1,162 @@
+        required: false
+        type: boolean
+        description: |
+          pull resistive up the pin. Can be used together with bias-pull-down.
```
```suggestion
          Enable pull-up resistor. Can be used together with bias-pull-down.
```

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -0,0 +1,162 @@
+        required: false
+        type: boolean
+        description: |
+          pull resistive down the pin. Can be used together with bias-pull-up.
```
```suggestion
          Enable pull-down resistor. Can be used together with bias-pull-up.
```

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -0,0 +1,162 @@
+      bias-pull-pin-default:
+        required: false
+        type: boolean
+        description: use pin-default pull state.
```
```suggestion
        description: use pin's default pull state
```

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -0,0 +1,162 @@
+      drive-strength:
+        required: false
+        type: int
+        description: sink or source at most X mA
```
```suggestion
        description: sink or source current in mA
```

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -0,0 +1,162 @@
+      drive-strength-microamp:
+        required: false
+        type: int
+        description: sink or source at most X uA
```
```suggestion
        description: sink or source current in μA
```

gmarull replied:
maybe `maximum sink or source current in μA` since it said `at most`.

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -0,0 +1,162 @@
+      drive-open-drain:
+        required: false
+        type: boolean
+        description: drive with open drain (hardware AND).
```
```suggestion
        description: drive with open drain (hardware AND)
```

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -0,0 +1,162 @@
+      drive-open-source:
+        required: false
+        type: boolean
+        description: drive with open source (hardware OR).
```
```suggestion
        description: drive with open source (hardware OR)
```

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -0,0 +1,162 @@
+        required: false
+        type: int
+        description: |
+          Takes the debounce time in usec as argument or 0 to disable debouncing
```
```suggestion
          debounce time in usec, or 0 to disable debouncing
```

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -0,0 +1,162 @@
+        required: false
+        type: int
+        description: |
+          this affects the expected clock skew on input pins and the delay
```
```suggestion
          This affects the expected clock skew on input pins and the delay
```

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -0,0 +1,162 @@
+      slew-rate:
+        required: false
+        type: int
+        description: set the slew rate
```
```suggestion
        description: slew rate in <INSERT_UNITS_HERE>
```

V/μs ?

gmarull replied:
At least for STM32 slew-rate is used as a kind of enum, see e.g. https://github.com/zephyrproject-rtos/hal_stm32/blob/master/dts/st/f4/stm32f401c(b-c)ux-pinctrl.dtsi#L111 It's weird, but Linux does the same (using `int` values 0-3)

mbolivar-nordic reviewed (changes requested):
Thanks @nandojve ! I think this is a great direction forward.

My comments are in 2 categories:

- docstring nits (see line suggestions)
- validation

It's not always clear what properties can be used together. In some cases, like `bias-pull-up` and `bias-pull-down`, it is, but other times I don't quite know. What if I put `bias-disable` together with `bias-high-impedance`? Legal?

If we are going to merge this with all these properties, I'd like to see a macro which lets us check that a node is defined correctly and isn't mixing anything it shouldn't be.

Basically something like this:

```
DT_PINCFG_CHECK(node_id)
```

This would use BUILD_ASSERT_MSG to check that the properties are defined in legal combinations only. If multiple BUILD_ASSERT_MSG statements are needed to print good error messages, that's OK too. Basically something like what I'm doing in https://github.com/zephyrproject-rtos/zephyr/pull/31965 with NRF_DT_PSEL_CHECK_EXACTLY_ONE() and NRF_DT_PSEL_CHECK_NOT_BOTH().

If that's too much work to write for all of these properties, maybe we can start small with just the ones we need, and build the CHECK macro up whenever we add new ones as necessary.

gmarull reviewed (commented):
I've tried with STM32 and I get the following:

```
devicetree error: /home/gmarull/ws/teslabs/zephyr/zephyr/dts/bindings/pinctrl/st,stm32-pinctrl.yaml (in 'slew-rate'): 'type' from included file overwritten ('int' replaced with 'string')
```

so `slew-rate` type can't be overriden as I suspected. @erwango should stm32 move to `int` so that we can follow Linux conventions? It will require minimal tweaks to the auto-generator.

nandojve commented:
Hello @mbolivar-nordic ,

I addressed nits as you suggested.

> * validation
> 
> It's not always clear what properties can be used together. In some cases, like `bias-pull-up` and `bias-pull-down`, it is, but other times I don't quite know. What if I put `bias-disable` together with `bias-high-impedance`? Legal?
>
> If we are going to merge this with all these properties, I'd like to see a macro which lets us check that a node is defined correctly and isn't mixing anything it shouldn't be.

You suggestion is a valid one and I really appreciated the proposal. However, I think it will be difficult combine in devicetree all platform rules because they can be divergent. The example that you pick is the most used and it can't be put in a rule, why? Atmel not allow **bias-pull-up** and **bias-pull-down** simultaneous however cypress admit and it is valid combination.

I think we should not try bring platform validation to devicetree because, eventually, we will face problems, like I just mention. At platform level, in general at drivers, validation must be check anyway because there are API's that can switch configurations, for instance. if it is not implemented now it can be in the near future. If we start to check the rules with macros may duplicate validation and bring more complexity to find bugs and errors over time.

What initially I've been working is populate bits to a variable that can be checked at pinctrl drivers somehow. If user add a property that platform don't support that is just ignored at build time. I mean, the property is not evaluated and no bit is set in the variable. These open possibility to populate devicetree with valid data even if not currently used.

I'll send another update to remove the _Can be used together with bias-pull-xx_ part to make it simpler to understand.

mbolivar-nordic reviewed (approved):
@nandojve thanks for the explanation. That seems reasonable for now. I'll try to think about ways to filter this a bit better in the tooling so that irrelevant properties can get removed.

erwango commented:
@galak, @mbolivar-nordic IIRC we discussed about a way to be picky in the yaml binding inclusion (via the use of allowlist, or excludelist).
Would it be doable/acceptable to exclude the slew-rate (int) property from pincfg-node.yaml inherit, and add a new slew-rate (enum) property in stm32-pinctrl.yaml ?

galak commented:
> @galak, @mbolivar-nordic IIRC we discussed about a way to be picky in the yaml binding inclusion (via the use of allowlist, or excludelist).
> Would it be doable/acceptable to exclude the slew-rate (int) property from pincfg-node.yaml inherit, and add a new slew-rate (enum) property in stm32-pinctrl.yaml ?

Oops, merged this before seeing @erwango comment.   Are we able to override the `slew-rate` in the stm32 to be an enum?

erwango commented:
> Are we able to override the slew-rate in the stm32 to be an enum?

Not as is.
The only way I see to support this is the above proposal if you and @mbolivar-nordic confirm this is doable/achievable.

galak commented:
> Not as is.
> The only way I see to support this is the above proposal if you and @mbolivar-nordic confirm this is doable/acceptable.

Yeah, re-opening my stale PR on that feature.
